src/workerd/io/observer.h

```diff
@@ -21,11 +21,24 @@ class WorkerInterface;
 class LimitEnforcer;
 class TimerChannel;
 
+class WebSocketObserver: public kj::Refcounted {
+public:
+  // Called when a worker sends a message on this WebSocket (includes close messages).
+  virtual void sentMessage(size_t bytes) { };
+  // Called when a worker receives a message is on this WebSocket (includes close messages).
+  virtual void receivedMessage(size_t bytes) { };
+};
+
 // Observes a specific request to a specific worker. Also observes outgoing subrequests.
 //
 // Observing anything is optional. Default implementations of all methods observe nothing.
 class RequestObserver: public kj::Refcounted {
 public:
+  // Called when a WebSocket terminating at a worker is coupled to the response. A WebSocketObserver
+  // may optionally be returned to observe messages being sent and received on the resulting
+  // websocket connection by the worker.
+  virtual kj::Maybe<kj::Own<WebSocketObserver>> coupledToLocalWebSocket() { return kj::none; };
```

Comment:
nit: why not simply `tryGetWebSocketObserver()` for the name?

Reply:
The idea behind the naming was that it is specifically called when the request is converted to an incoming websocket connection which terminates in a "local" websocket i.e. in a worker. `tryGetWebSocketObserver()` doesn't indicate any of those circumstances, but I guess I could just document the above rather than trying to encode it into the name.